File: app/src/main/java/com/unitpricecalculator/unit/UnitEntry.java
package com.unitpricecalculator.unit;

import com.google.common.base.Preconditions;

public final class UnitEntry {

    private final double cost;

    private final String costString;

    private final int quantity;

    private final String quantityString;

    private final double size;

    private final String sizeString;

    private final Unit unit;

    private UnitEntry(double cost, String costString, int quantity, String quantityString, double size,
                      String sizeString, Unit unit) {
        Preconditions.checkArgument(cost >= 0);
        Preconditions.checkArgument(quantity > 0);
        Preconditions.checkArgument(size > 0);
        this.cost = cost;
        this.costString = Preconditions.checkNotNull(costString);
        this.quantity = quantity;
        this.quantityString = Preconditions.checkNotNull(quantityString);
        this.size = size;
        this.sizeString = Preconditions.checkNotNull(sizeString);
        this.unit = Preconditions.checkNotNull(unit);
    }

    public double getCost() {
        return cost;
    }

    public String getCostString() {
        return costString;
    }

    public int getQuantity() {
        return quantity;
    }

    public String getQuantityString() {
        return quantityString;
    }

    public double getSize() {
        return size;
    }

    public String getSizeString() {
        return sizeString;
    }

    public Unit getUnit() {
        return unit;
    }

    public double getPricePerUnit() {
        return cost / (quantity * size);
    }

    public double pricePer(double size, Unit unit) {
        Preconditions.checkArgument(unit.getUnitType() == this.unit.getUnitType());
        double costPerUnit = getPricePerUnit();
        double costPerOtherUnit = costPerUnit / (this.unit.getFactor() / unit.getFactor());
        return size * costPerOtherUnit;
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private double cost;

        private String costString;

        private int quantity;

        private String quantityString;

        private double size;

        private String sizeString;

        private Unit unit;

        private Builder() {
            this.quantity = 1;
        }

        public Builder setCost(double cost) {
            Preconditions.checkArgument(cost >= 0);
            this.cost = cost;
            return this;
        }

        public Builder setQuantity(int quantity) {
            Preconditions.checkArgument(quantity > 0);
            this.quantity = quantity;
            return this;
        }

        public Builder setUnit(Unit unit) {
            this.unit = Preconditions.checkNotNull(unit);
            return this;
        }

        public Builder setSize(double size) {
            Preconditions.checkArgument(size > 0);
            this.size = size;
            return this;
        }

        public Builder setCostString(String costString) {
            this.costString = Preconditions.checkNotNull(costString);
            return this;
        }

        public Builder setQuantityString(String quantityString) {
            this.quantityString = Preconditions.checkNotNull(quantityString);
            return this;
        }

        public Builder setSizeString(String sizeString) {
            this.sizeString = Preconditions.checkNotNull(sizeString);
            return this;
        }

        public UnitEntry build() {
            return new UnitEntry(cost, costString, quantity, quantityString, size, sizeString, unit);
        }

        @Override
        public String toString() {
            return super.toString();
        }
    }
}

File: app/src/main/java/com/unitpricecalculator/unit/UnitEntry.kt
package com.unitpricecalculator.unit

data class UnitEntry constructor(
    val cost: Double,
    val costString: String,
    val quantity: Int,
    val quantityString: String?,
    val size: Double,
    val sizeString: String,
    val unit: Unit
) {
    init {
        require(cost >= 0)
        require(quantity > 0)
        require(size > 0)
    }

    private val pricePerUnit: Double
        get() = cost / (quantity * size)
    val pricePerBaseUnit: Double
        get() = pricePer(1.0, unit.unitType.base)

    fun pricePer(size: Double, unit: Unit): Double {
        require(unit.unitType == this.unit.unitType) {
            "Expected " + this.unit.unitType + ", was " + unit.unitType
        }
        val costPerUnit = pricePerUnit
        val costPerOtherUnit = costPerUnit / (this.unit.factor / unit.factor)
        return size * costPerOtherUnit
    }

    class Builder {
        private var cost = 0.0
        private var costString: String? = null
        private var quantity = 1
        private var quantityString: String? = null
        private var size = 0.0
        private var sizeString: String? = null
        private var unit: Unit? = null

        fun setCost(cost: Double): Builder {
            require(cost >= 0)
            this.cost = cost
            return this
        }

        fun setQuantity(quantity: Int): Builder {
            require(quantity > 0)
            this.quantity = quantity
            return this
        }

        fun setUnit(unit: Unit): Builder {
            this.unit = unit
            return this
        }

        fun setSize(size: Double): Builder {
            require(size > 0)
            this.size = size
            return this
        }

        fun setCostString(costString: String): Builder {
            this.costString = costString
            return this
        }

        fun setQuantityString(quantityString: String): Builder {
            this.quantityString = quantityString
            return this
        }

        fun setSizeString(sizeString: String): Builder {
            this.sizeString = sizeString
            return this
        }

        fun build(): UnitEntry {
            return UnitEntry(
                cost,
                checkNotNull(costString) { "costString" },
                quantity,
                quantityString,
                size,
                checkNotNull(sizeString) { "sizeString" },
                checkNotNull(unit) { "unit" }
            )
        }
    }
}
